File: streamable/python-mcp-streamable-http/src/tools.py
"""
Pure tool logic for the demo MCP server.

These functions hold the actual business logic of each MCP tool and are
intentionally kept transport-agnostic. This makes them trivial to unit-test
without spinning up the HTTP server.

Server name, version and supported tools are exposed as constants so that
`server.py`, `client.py`, tests and the `app://info` resource can all share a
single source of truth.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

SERVER_NAME = "demo-python-streamable-http"
SERVER_VERSION = "0.1.0"
SERVER_LANGUAGE = "python"
SERVER_TRANSPORT = "streamable-http"

TOOL_NAMES: list[str] = [
    "echo",
    "add_numbers",
    "get_server_time",
    "generate_demo_log",
    "get_system_status",
]

RESOURCE_URIS: list[str] = ["app://info"]

PROMPT_NAMES: list[str] = ["summarize_text", "explain_code"]


# Server start time, used by `get_system_status` to compute uptime.
_SERVER_START_MONOTONIC = time.monotonic()


def echo(text: str) -> str:
    """Return the exact same text. Useful for verifying connectivity."""
    if not isinstance(text, str):
        raise ValueError("`text` must be a string")
    return text


def add_numbers(a: float, b: float) -> float:
    """Add two numbers and return the result.

    Both arguments must be numeric (int or float). Booleans are rejected
    because in Python `bool` is a subclass of `int` but allowing it would be
    surprising to MCP callers.
    """
    for name, value in (("a", a), ("b", b)):
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            raise ValueError(f"`{name}` must be a number, got {type(value).__name__}")
    return float(a) + float(b)


def get_server_time() -> dict[str, str]:
    """Return the current server time in ISO 8601 format, with timezone info."""
    # `astimezone()` without args uses the system local timezone.
    now_local = datetime.now().astimezone()
    return {
        "iso": now_local.isoformat(),
        "utc": datetime.now(timezone.utc).isoformat(),
        "timezone": str(now_local.tzinfo),
    }


@dataclass
class DemoLogStep:
    step: int
    total: int
    message: str
    timestamp: str


def build_demo_log_steps(steps: int) -> list[DemoLogStep]:
    """Build the list of log entries that `generate_demo_log` would emit.

    Pure function so it can be unit-tested without involving the server.
    """
    if not isinstance(steps, int) or isinstance(steps, bool):
        raise ValueError("`steps` must be an integer")
    if steps < 1 or steps > 50:
        raise ValueError("`steps` must be between 1 and 50")

    entries: list[DemoLogStep] = []
    for i in range(1, steps + 1):
        entries.append(
            DemoLogStep(
                step=i,
                total=steps,
                message=f"Processing step {i}/{steps}",
                timestamp=datetime.now(timezone.utc).isoformat(),
            )
        )
    return entries


def get_system_status() -> dict[str, Any]:
    """Return a small dictionary describing the server's current state."""
    uptime_seconds = round(time.monotonic() - _SERVER_START_MONOTONIC, 3)
    return {
        "language": SERVER_LANGUAGE,
        "name": SERVER_NAME,
        "version": SERVER_VERSION,
        "transport": SERVER_TRANSPORT,
        "uptime_seconds": uptime_seconds,
        "tools": list(TOOL_NAMES),
        "resources": list(RESOURCE_URIS),
        "prompts": list(PROMPT_NAMES),
    }


def app_info(endpoint_path: str = "/mcp") -> dict[str, Any]:
    """Payload returned by the `app://info` resource."""
    return {
        "name": SERVER_NAME,
        "language": SERVER_LANGUAGE,
        "version": SERVER_VERSION,
        "transport": SERVER_TRANSPORT,
        "endpoint": endpoint_path,
        "tools": list(TOOL_NAMES),
        "resources": list(RESOURCE_URIS),
        "prompts": list(PROMPT_NAMES),
    }


def summarize_text_prompt(text: str) -> str:
    """Reusable prompt asking the model to summarize a piece of text."""
    if not isinstance(text, str) or not text.strip():
        raise ValueError("`text` must be a non-empty string")
    return (
        "Please produce a concise, faithful summary of the text below.\n"
        "Keep it under 5 sentences, preserve the original meaning, and do not "
        "invent facts that are not present in the text.\n\n"
        "---BEGIN TEXT---\n"
        f"{text}\n"
        "---END TEXT---"
    )


def explain_code_prompt(code: str, language: str = "python") -> str:
    """Reusable prompt asking the model to explain a code snippet."""
    if not isinstance(code, str) or not code.strip():
        raise ValueError("`code` must be a non-empty string")
    if not isinstance(language, str) or not language.strip():
        language = "unknown"
    return (
        f"Explain the following {language} code as if you were teaching a "
        "junior engineer. Cover:\n"
        "1. What the code does at a high level.\n"
        "2. The role of each major construct or function.\n"
        "3. Any non-obvious behavior, edge cases or potential bugs.\n\n"
        f"```{language}\n{code}\n```"
    )

File: streamable/typescript-mcp-streamable-http/src/tools.ts
/**
 * Pure tool logic for the demo MCP server.
 *
 * These functions hold the actual business logic of each tool, kept
 * transport-agnostic so they can be unit-tested without spinning up the
 * HTTP server. `server.ts` only wires them into MCP.
 */

export const SERVER_NAME = "demo-typescript-streamable-http";
export const SERVER_VERSION = "0.1.0";
export const SERVER_LANGUAGE = "typescript";
export const SERVER_TRANSPORT = "streamable-http";

export const TOOL_NAMES = [
  "echo",
  "add_numbers",
  "get_server_time",
  "generate_demo_log",
  "get_system_status",
] as const;

export const RESOURCE_URIS = ["app://info"] as const;

export const PROMPT_NAMES = ["summarize_text", "explain_code"] as const;

// Captured at module-load time so `get_system_status` can compute uptime.
const SERVER_START_MS = Date.now();

export function echo(text: string): string {
  if (typeof text !== "string") {
    throw new Error("`text` must be a string");
  }
  return text;
}

export function addNumbers(a: number, b: number): number {
  for (const [name, value] of [
    ["a", a],
    ["b", b],
  ] as const) {
    if (typeof value !== "number" || Number.isNaN(value) || !Number.isFinite(value)) {
      throw new Error(`\`${name}\` must be a finite number`);
    }
  }
  return a + b;
}

export interface ServerTime {
  iso: string;
  utc: string;
  timezone: string;
}

export function getServerTime(): ServerTime {
  const now = new Date();
  // Best-effort detection of the host's IANA timezone (e.g. "Europe/Madrid").
  let timezone = "UTC";
  try {
    timezone = Intl.DateTimeFormat().resolvedOptions().timeZone ?? "UTC";
  } catch {
    timezone = "UTC";
  }
  return {
    iso: now.toISOString(),
    utc: now.toISOString(),
    timezone,
  };
}

export interface DemoLogStep {
  step: number;
  total: number;
  message: string;
  timestamp: string;
}

export function buildDemoLogSteps(steps: number): DemoLogStep[] {
  if (!Number.isInteger(steps)) {
    throw new Error("`steps` must be an integer");
  }
  if (steps < 1 || steps > 50) {
    throw new Error("`steps` must be between 1 and 50");
  }
  const out: DemoLogStep[] = [];
  for (let i = 1; i <= steps; i++) {
    out.push({
      step: i,
      total: steps,
      message: `Processing step ${i}/${steps}`,
      timestamp: new Date().toISOString(),
    });
  }
  return out;
}

export interface SystemStatus {
  language: string;
  name: string;
  version: string;
  transport: string;
  uptime_seconds: number;
  tools: string[];
  resources: string[];
  prompts: string[];
}

export function getSystemStatus(): SystemStatus {
  const uptimeSeconds = Math.round(((Date.now() - SERVER_START_MS) / 1000) * 1000) / 1000;
  return {
    language: SERVER_LANGUAGE,
    name: SERVER_NAME,
    version: SERVER_VERSION,
    transport: SERVER_TRANSPORT,
    uptime_seconds: uptimeSeconds,
    tools: [...TOOL_NAMES],
    resources: [...RESOURCE_URIS],
    prompts: [...PROMPT_NAMES],
  };
}

export interface AppInfo {
  name: string;
  language: string;
  version: string;
  transport: string;
  endpoint: string;
  tools: string[];
  resources: string[];
  prompts: string[];
}

export function appInfo(endpointPath: string = "/mcp"): AppInfo {
  return {
    name: SERVER_NAME,
    language: SERVER_LANGUAGE,
    version: SERVER_VERSION,
    transport: SERVER_TRANSPORT,
    endpoint: endpointPath,
    tools: [...TOOL_NAMES],
    resources: [...RESOURCE_URIS],
    prompts: [...PROMPT_NAMES],
  };
}

export function summarizeTextPrompt(text: string): string {
  if (typeof text !== "string" || text.trim() === "") {
    throw new Error("`text` must be a non-empty string");
  }
  return (
    "Please produce a concise, faithful summary of the text below.\n" +
    "Keep it under 5 sentences, preserve the original meaning, and do not " +
    "invent facts that are not present in the text.\n\n" +
    "---BEGIN TEXT---\n" +
    `${text}\n` +
    "---END TEXT---"
  );
}

export function explainCodePrompt(code: string, language: string = "typescript"): string {
  if (typeof code !== "string" || code.trim() === "") {
    throw new Error("`code` must be a non-empty string");
  }
  const lang = typeof language === "string" && language.trim() !== "" ? language : "unknown";
  return (
    `Explain the following ${lang} code as if you were teaching a ` +
    "junior engineer. Cover:\n" +
    "1. What the code does at a high level.\n" +
    "2. The role of each major construct or function.\n" +
    "3. Any non-obvious behavior, edge cases or potential bugs.\n\n" +
    `\`\`\`${lang}\n${code}\n\`\`\``
  );
}
